packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -791,30 +797,36 @@ export interface SimplifyValueNode<T> {
 export interface SimplifyContext<T> {
 	node: SimplifyValueNode<T>
 	ctx: McdocCheckerContext<T>
-	structFields?: {
-		literalFields: Map<string, StructTypePairField>
-		complexFields: SimplifiedStructTypePairField[]
-	}
 	isMember?: boolean
 	typeArgs?: SimplifiedMcdocType[]
 	typeMapping?: { [path: string]: SimplifiedMcdocType }
 }
+export interface SimplifyResult<T extends SimplifiedMcdocType> {
+	typeDef: T
+	dynamicData?: boolean
```

Comment:
Not a fan of a special return type with a magic flag that enables/disables caching. This indicates to me that this is caching stuff on the wrong level. From what I can tell, if a single entity suddenly uses a dynamic index somewhere, it completely disables caching for the entity fallback case.

Reply:
> From what I can tell, if a single entity suddenly uses a dynamic index somewhere, it completely disables caching for the entity fallback case.

Well, maybe yes? I am not sure you mean the correct thing here. Most of the result will still be cached. Caching something that requires a dynamic index is impossible (or you would need to cache the runtime value somehow, which I don't think would improve the performance, as the static index is still cached as I will explain below).

```rs
struct Llama {
    ...Breedable
    DecorItem?: InventoryItem
}

struct Breedable {
	...super::MobBase,
	/// Ticks until it stops searching for a mate.
	InLove?: int @ 0..,
	Age?: int,
	/// A value of age assigned to this mob when it grows up.
	/// Incremented when a baby mob is fed.
	ForcedAge?: int,
	/// Lower bits of the UUID of the player that caused this mob to breed.
	#[until="1.16"]
	LoveCauseLeast?: long,
	/// Upper bits of the UUID of the player that caused this mob to breed.
	#[until="1.16"]
	LoveCauseMost?: long,
	/// Player that caused this mob to breed.
	#[since="1.16"] #[uuid]
	LoveCause?: int[] @ 4,
}

struct Creeper {
	...super::MobBase,
	/// Whether it is being struck by lightning.
	powered?: boolean,
	/// Radius of the explosion.
	ExplosionRadius?: byte,
	/// Ticks until it explodes.
	Fuse?: short,
	/// Whether it was lit with flint and steel.
	ignited?: boolean,
}

struct MobBase {
	/// [main hand, offhand]
	HandItems?: [(InventoryItem | struct {})] @ 2,
	/// [feet, legs, body, head]
	ArmorItems?: [(InventoryItem | struct {})] @ 4,
}

dispatch minecraft:entity[creeper] to Creeper
dispatch minecraft:entity[llama] to Llama
```

Assuming we resolved the fallback case once, this will result in the following cached elements:
```md
- entity dispatcher
  - creeper
  - llama
- ::MobBase
- ::Breedable
- ::Creeper
- ::Llama
```

Now, let's say Llama was for some reason ammended with `...mincraft:spitting_behavior[[spitting_type]], spitting_type: string`

Now the following would be cached:
```md
- entity dispatcher
  - creeper
- ::MobBase
- ::Breedable
- ::Creeper
```

And when resolving the fallback dispatcher, llama would never hit the cache, never be cached and always resolved anew, however when resolving Llama, it would only do that, and not e.g. `::Breedable`.

Also note that this definition does not influence caching, since simplify is not deep:
```rs
struct {
    spitting_type: string,
    spitting_data: mincraft:spitting_behavior[[spitting_type]],
}
```

Also, this would be a hit after resolving fallback once with runtime value `{entity_id: creeper}`:
```rs
struct {
    entity_id: string
    entity_data: minecraft:entity[[entity_id]]
}
```
Basically, it would do the dynamic lookup, find creeper, then lookup creeper in the dispatcher and find the cached simplified value for creeper. This works, because the dynamic dispatcher `minecraft:entity[[entity_id]]` is not cached, but after resolving the dynamic indices, the result of the static members is cached.

> Not a fan of a special return type with a magic flag that enables/disables caching. This indicates to me that this is caching stuff on the wrong level.

Could you elaborate on this? I don't really agree, or understand that concern. Caching dynamic looksups is impossible by definition.

Reply:
Hmm okay, I've slowly been realizing that maybe this is the best way to cache...